app/frontend/actions.js

```diff
@@ -864,6 +937,70 @@ module.exports = ({setState, getState}) => {
     validateSendFormAndCalculateFee()
   }
 
+  const updateStakePoolAddress = (state, e) => {
+    setState({
+      stakePools: state.stakePools.map((pool) => {
+        return Object.assign(
+          {},
+          pool,
+          pool.ID === parseInt(e.target.id, 10)
+            ? {
+              address: e.target.value,
+              ...fetchPoolInfo(e.target.value),
+            }
+            : {}
+        )
+      }),
+    })
+    validateDelegationAndCalculateDelegationFee()
+  }
+
+  const updateStakePoolPercent = (state, e) => {
+    const ID = parseInt(e.target.id, 10)
+    const oldPercent = state.stakePools
+      .map((pool) => (pool.ID === ID ? pool.percent : 0))
+      .reduce((x, y) => x + y, 0)
+    const nextPercent = Math.min(
+      100 - state.delegatedPercent + oldPercent,
+      parseInt(e.target.value ? e.target.value : 0, 10)
+    )
+
+    setState({
+      delegatedPercent: state.delegatedPercent - oldPercent + nextPercent,
+      stakePools: state.stakePools.map((pool) => {
+        return Object.assign({}, pool, pool.ID === ID ? {percent: nextPercent} : {})
+      }),
+    })
+    validateDelegationAndCalculateDelegationFee()
+  }
+
+  const addStakePool = (state, e) => {
+    setState({
+      stakePools: state.stakePools.concat({
+        address: '',
+        percent: 0,
+        name: '',
+        valid: false,
+        ID: state.stakePoolID + 1,
```

Comment:
This is definitely not what we want. You don't need to store index. Just get it when iterating: `stakePools.map((data, i) => ....)`

Reply:
This was rubbish indeed, I wasn't aware there exists indexed iteration. Fixed and thanks.